src/ProjectSystemTools/BuildLogging/Model/Backend/BackendBuildTableDataSource.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+using Microsoft.Build.Evaluation;
+using Microsoft.Build.Framework;
+using Microsoft.VisualStudio.ProjectSystem.Tools.Providers;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Tools.BuildLogging.Model.BackEnd
+{
+    [Export(typeof(ILoggingController))]
+    [Export(typeof(ILoggingDataSource))]
+    internal sealed class BackEndBuildTableDataSource : ILoggingController, ILoggingDataSource
+    {
+        private const string BuildDataSourceDisplayName = "Build Data Source";
+        private const string BuildTableDataSourceIdentifier = nameof(BuildTableDataSourceIdentifier);
+        private const string BuildTableDataSourceSourceTypeIdentifier = nameof(BuildTableDataSourceSourceTypeIdentifier);
+
+        private readonly EvaluationLogger _evaluationLogger;
+        private readonly RoslynLogger _roslynLogger;
+
+        private ImmutableList<Build> _entries = ImmutableList<Build>.Empty;
+
+        public string SourceTypeIdentifier => BuildTableDataSourceSourceTypeIdentifier;
+
+        public string Identifier => BuildTableDataSourceIdentifier;
+
+        public string DisplayName => BuildDataSourceDisplayName;
+
+        public bool SupportRoslynLogging => _roslynLogger.Supported;
+
+        public bool IsLogging { get; private set; }
+
+        public bool SupportsRoslynLogging => _roslynLogger.Supported;
+
+        private Action NotifyUI { get; set; }
+
+        public BackEndBuildTableDataSource()
+        {
+            _evaluationLogger = new EvaluationLogger(this);
+            _roslynLogger = new RoslynLogger(this);
+        }
+
+        public void Start(Action notifyCallback)
+        {
+            NotifyUI = notifyCallback;
+
+            IsLogging = true;
+            ProjectCollection.GlobalProjectCollection.RegisterLogger(_evaluationLogger);
+            _roslynLogger.Start();
+        }
+
+        public void Stop()
+        {
+            NotifyUI = null;
+
+            IsLogging = false;
+            ProjectCollection.GlobalProjectCollection.UnregisterAllLoggers();
+            _roslynLogger.Stop();
+        }
+
+        public void Clear()
+        {
+            foreach (var build in _entries)
+            {
+                build.Dispose();
+            }
+            _entries = ImmutableList<Build>.Empty;
+        }
+
+        public ILogger CreateLogger(bool isDesignTime) => new ProjectLogger(this, isDesignTime);
+
+        /// <summary>
+        /// return log path on server for a given build
+        /// If buildID cannot be found, will return null
```

Comment:
> If buildID cannot be found, will return null

Is this true? It looks like it'll throw `NullReferenceException` if the build is not found.

Reply:
Ah, yeah I wrote that documentation weirdly 😄